enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -348,78 +377,58 @@ impl<'a> Handler<'a> {
         self.trace("readv", 3);
 
         let fd = self.aex.gpr.rdi.raw();
-        let trusted = unsafe {
+        let tiovec = unsafe {
             from_raw_parts_mut(
                 self.aex.gpr.rsi.raw() as *mut libc::iovec,
                 self.aex.gpr.rdx.raw() as usize,
             )
         };
 
-        // Add up total size of buffers and size of iovec array.
-        let bufsize = trusted.iter().fold(0, |a, e| a + e.iov_len);
-        let iovecsize = size_of::<libc::iovec>() * trusted.len();
-        let size = bufsize + iovecsize;
-
-        // Allocate some unencrypted memory.
-        let map = match self.ualloc(size as u64) {
-            Err(errno) => return errno as u64,
-            Ok(map) => unsafe { from_raw_parts_mut(map, size as usize) },
-        };
-
-        // Split allocated memory into that used by the iovec struct array and that used by its buffers.
-        let (uiovec, ubuffer) = map.split_at_mut(iovecsize);
+        // Allocate the correct number of iovecs in untrusted memory.
+        let mut c = Cursor::new(&mut self.block.buf);
+        let uiovec: &mut [libc::iovec] = c.alloc(tiovec.len()).unwrap();
 
-        // Convert the prefix from a byte slice into an iovec slice.
-        let (_, untrusted, _) = unsafe { uiovec.align_to_mut::<libc::iovec>() };
-        if untrusted.len() != trusted.len() {
-            self.attacked();
+        // Update pointers from iovecs to buffers in untrusted memory.
+        let mut size = 0;
+        for (t, u) in tiovec.iter().zip(uiovec.iter_mut()) {
+            size += t.iov_len;
+            u.iov_base = c.alloc(t.iov_len).unwrap().as_mut_ptr();
```

Comment:
```suggestion
            u.iov_len = t.iov_len;
            u.iov_base = c.alloc(t.iov_len).unwrap().as_mut_ptr();
```

Don't forget to initialize the length!

Reply:
Good catch, thanks.